---
layout: page
title: Why use dDocent?
subtitle: 
---


# Accuracy



## *de nono* assembly

dDocent employs a series of data reduction techniques, aligment based clustering (using [CD-hit](https://github.com/weizhongli/cdhit)), and, for PE assembly, a specialized RAD assembly software ([rainbow](https://sourceforge.net/projects/bio-rainbow/files/).  This combination allows for accurate and effecient *de novo* assembly.  

### A comparison among pipelines

![alt tag](https://raw.githubusercontent.com/jpuritz/dDocent/gh-pages/RADassembly.png)

This is 1000 simulated ddRAD data loci being assembled across a variety of parameters for each pipeline.

### Overclustering leads to bias

![alt tag](https://raw.githubusercontent.com/jpuritz/dDocent/master/Sample%20Comparsion.png)

Above is a figure depicting the relative bias of pairwise FST values generated by different RADseq bioinformatic pipelines. Circles are sized according to the magnitude of bias (Observed - Expected)/Expected and are colored relative to the percentage of INDEL variation simulated in the data set: blue- 1%, red-5%, and green-10%.  Simulations consisted of four populations in a stepping stone model with a decreasing migration rates.  

# Bayesian, haplotype based, population-aware, genotyping from FreeBayes

[*FreeBayes*](http://arxiv.org/abs/1207.3907) is a 
[Bayesian](http://en.wikipedia.org/wiki/Bayesian_inference) genetic variant 
detector designed to detect  SNPs, INDels (insertions and deletions), and complex events (composite insertion and 
substitution events) smaller than the length of a short-read sequencing 
alignment.  *FreeBayes* is haplotype-based, in the sense that it calls variants based on 
the literal sequences of reads aligned to a particular target, not their 
precise alignment, and  for any number of individuals from a population and a 
to determine the most-likely combination of genotypes for the population at 
each position in the reference.  


